api/src/main/java/jakarta/json/bind/decorator/BuilderProviderLoader.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2021 Oracle and/or its affiliates. All rights reserved.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0, which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception, which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ */
+
+package jakarta.json.bind.decorator;
+
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.ServiceLoader;
+
+import jakarta.json.bind.spi.DecoratorBuilderProvider;
+
+/**
+ * SPI loader of the {@link DecoratorBuilderProvider}.
+ */
+class BuilderProviderLoader {
+
+    private static DecoratorBuilderProvider instance = null;
+
+    private BuilderProviderLoader() {
+        throw new IllegalStateException("This class cannot be instantiated");
+    }
+
+    static DecoratorBuilderProvider provider() {
+        if (instance == null) {
+            synchronized (DecoratorBuilderProvider.class) {
+                if (instance != null) {
+                    return instance;
+                }
+                PrivilegedAction<DecoratorBuilderProvider> loader = () ->
```

Comment:
Well, I am not sure I agree here, but lets assume I am changing that. So do you think, I should drop the BuilderProviderLoader and change that DecoratorBuilderProvider to not be the SPI, but add new method to the JsonbProvider which would create this DecoratorBuilderProvider impl?

Reply:
There are two points:

1. Do we want another entry point than JsonbBuilder? (guess not)
2. Ensure we have a single call to ServiceLoader in the whole module. (i doubt very less on this one ;))

Reply:
I gave this some thought. I think I am fine with moving those decorator builder methods from `DecoratorBuilderProvider` to `JsonbProvider`, if that was what you want me to do :-) By this change I will completely drop the loader and decorator provider. But I will also need to change how is `JsonbProvider` currently loaded since it is not cached and it is done every single `provider()` method call.

Reply:
Fine with me

Side note being: no impacted change since spec code already has this pitfall and does not work correctly in all apps (= not  a drawback ;))

Reply:
Alright, both providers merged :-) Now I will just change that provider method impl.

Reply:
SPI discovery changed. What to you think now? :-)